processing/src/main/java/org/apache/druid/math/expr/vector/FallbackVectorProcessor.java

```diff
@@ -0,0 +1,422 @@
+     */
+    public void populate(final Expr.VectorInputBinding vectorBindings)
+    {
+      this.results = processor.evalVector(vectorBindings);
+    }
+
+    /**
+     * Set {@link #rowNum}, which controls which row of {@link #results} is returned by {@link #eval(ObjectBinding)}.
+     */
+    public void setRowNumber(final int rowNum)
+    {
+      this.rowNum = rowNum;
+    }
+
+    @Override
+    public ExprEval eval(ObjectBinding bindings)
+    {
+      if (results == null) {
+        // "results" can be null if eval is called during ExprMacro#apply.
+        return originalExpr.eval(bindings);
+      }
+
+      // In all other cases, ignore the provided bindings and use the computed "results" instead.
+      if (type.is(ExprType.LONG)) {
+        final boolean isNull = results.getNullVector() != null && results.getNullVector()[rowNum];
+        return ExprEval.ofLong(isNull ? null : results.getLongVector()[rowNum]);
+      } else if (type.is(ExprType.DOUBLE)) {
+        final boolean isNull = results.getNullVector() != null && results.getNullVector()[rowNum];
+        return ExprEval.ofDouble(isNull ? null : results.getDoubleVector()[rowNum]);
+      } else {
+        return ExprEval.ofType(type, results.getObjectVector()[rowNum]);
+      }
+    }
+
+    @Override
+    public String stringify()
+    {
+      throw DruidException.defensive(
+          "Unexpected call to stringify in fallback processor for expr[%s]",
+          originalExpr.stringify()
+      );
+    }
+
+    @Override
+    public Expr visit(Shuttle shuttle)
+    {
+      throw DruidException.defensive(
+          "Unexpected call to visit in fallback processor for expr[%s]",
+          originalExpr.stringify()
+      );
+    }
+
+    @Override
+    public BindingAnalysis analyzeInputs()
+    {
+      return originalExpr.analyzeInputs();
+    }
+
+    @Override
+    public boolean isLiteral()
+    {
+      return originalExpr.isLiteral();
+    }
+
+    @Override
+    public boolean isNullLiteral()
+    {
+      return originalExpr.isNullLiteral();
+    }
+
+    @Nullable
+    @Override
+    public Object getLiteralValue()
+    {
+      return originalExpr.getLiteralValue();
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      AdaptedExpr that = (AdaptedExpr) o;
+      return Objects.equals(originalExpr, that.originalExpr) && Objects.equals(type, that.type);
+    }
+
+    @Override
+    public int hashCode()
+    {
+      return Objects.hash(originalExpr, type);
+    }
+  }
+
+  /**
+   * Implementation of {@link Expr.ObjectBinding} where we do not actually expect any methods to be called. This is
+   * because bindings should only be used by identifiers, and this fallback processor is never used to
+   * implement identifiers.
+   */
+  private static class UnusedBinding implements Expr.ObjectBinding
```

Comment:
nit: this seems pretty similar to the thing made by `InputBindings.validateConstant` which is used during sql planning to reduce constant expressions, maybe should consolidate.

I guess not totally possible though since the other one attaches the expression using the constant bindings for nicer validation messages, but that's not a thing here so it can be static and doesn't have the need since it should only be used for identifiers

Reply:
Yeah, they are very similar, but probably makes sense to keep them different for error-message reasons.